introduce a new Consumer pattern

mmmries opened the pull request:
This is a proposal to provide an alternative to the current pattern of the `Fireworks.Channel` subscriber.

__The Use Case__

I have an application that will need to subscriber to 38 queues. 33 of the queues are just triplets of `created`, `updated` and `deleted` queues for a given resource type. Rather than write (or generate) 38 unique modules in my project I would like to setup a supervisor that supervises 38 consumer processes each of which are subscribed to a given queue and consume the messages by calling a function like `MyApp.BudgetSubscriber.created`.

<details>
  <summary>Screenshot showing queue pattern</summary>
  <img src="https://cloud.githubusercontent.com/assets/80008/24879528/77fedd92-1df4-11e7-9398-b8d94dcef3c6.png" rel="screenshot showing queues" />
</details>

__The Implementation__

This is mostly a re-write of the `Fireworks.Channel` pattern, except that it replaces the module-level callbacks with functions that can be passed in when standing up a supervision tree.

This also changes a few other behaviors:
* `ack` and `reject` are now synchronous using `GenServer.call` instead of `GenServer.cast`. This way if you try to ack a message twice you'll get an error in the consuming code.
* `ack` and `reject` are called with a full copy of the `meta` parameter rather than pulling the delivery tag out of it. This avoids the problem where after a connection recovery someone tries to ack a message that was received from the old connection/channel. RabbitMQ will forcibly shut down your channel if you try to ack a delivery tag that hasn't been given to your channel, or even worse you might successfully `ack` a delivery tag that has been received under the new connection, causing a very confusing double-ack problem. To avoid this we send back both the `channel` and the `delivery_tag` so we can return an error if someone tries to send back an `ack` or `reject` for a message received under an old channel.
* removed the `publish` function which was just a proxy to `Fireworks.publish`
* removed the automatic JSON decoding. I can definitely add this back in, but it looks like there have been a lot of reported issues around this feature. Maybe it's best to let client code handle the encoding/decoding and we just handle availability and delivery of messages?

__Usage__

In my application I would use this by doing something like:

```elixir
children = [
  worker(Fireworks.Consumer, [%{
    queue: "brokaw.abacus.budget.created",
    prefetch: 2,
    consume_fn: &Brokaw.BudgetSubscriber.created/2,
    setup_fn: fn(channel) ->
      Queue.declare(channel, "brokaw.abacus.budget.created", durable: false)
      Queue.bind(channel, "brokaw.abacus.budget.created", "events", routing_key: "abacus.budget.created")
    end},
    [name: :budget_created]],
    [id: :budget_created])
```

That's a bit of a mouthful, but it lets me pass in all of my parameters as data so I can create helper functions to generate most of that data and easily test those functions.

Is this a pattern that looks palatable to the `fireworks` community? For my application this makes it easy for me to generate all 38 subscriptions with ~20 lines of code since they follow a naming convention and I can easily do all the configuration at runtime instead of splitting it across my module and my config files. I realize of course that my use-case is specific to my company/application so I can totally understand if this isn't a direction that `fireworks` wants to go.

/cc @ewitchin

mmmries committed:
introduce a new Consumer pattern that can be supervised by client applications

mmmries commented:
Oh, forgot to mention that I'd be happy to add documentation and tests around the `Consumer` if this is a direction that looks promising.

mmmries committed:
refactor to avoid compiler warning

mmmries committed:
add some error reporting and only reject messages that failed for the current connection

ewitchin commented:
@mmmries I will take a look at this a play around with it probably a little later this week. Thanks for working on this!

mmmries commented:
Anyone had a chance to look at this? I'm getting close to rolling a first version of an app using this pattern to production and I have a supervisor that currently looks like this:

<details>
  <summary>lib/brokaw/subscriber.ex</summary>
  
```elixir
defmodule Brokaw.Subscriber do
  use Supervisor
  alias Brokaw.Subscriber.{
    Account,
    Budget,
    Category,
    Device,
    DeliverableNotification,
    Goal,
    Holding,
    Member,
  }

  def start_link do
    Supervisor.start_link(__MODULE__, [])
  end

  def init([]) do
    true = Process.register(self(), Brokaw.Subscriber)
    children = [
      subscriber_worker("abacus.account.created", Account, :created),
      subscriber_worker("abacus.account.deleted", Account, :deleted),
      subscriber_worker("abacus.account.updated", Account, :updated),
      subscriber_worker("abacus.budget.created", Budget, :created),
      subscriber_worker("abacus.budget.deleted", Budget, :deleted),
      subscriber_worker("abacus.budget.updated", Budget, :updated),
      subscriber_worker("abacus.category.created", Category, :created),
      subscriber_worker("abacus.category.deleted", Category, :deleted),
      subscriber_worker("abacus.category.updated", Category, :updated),
      subscriber_worker("amigo.device.created", Device, :created),
      subscriber_worker("amigo.device.deleted", Device, :deleted),
      subscriber_worker("amigo.device.updated", Device, :updated),
      subscriber_worker("newman.deliverable_notification.created", DeliverableNotification, :created),
      subscriber_worker("newman.deliverable_notification.updated", DeliverableNotification, :updated),
      subscriber_worker("abacus.goal.created", Goal, :created),
      subscriber_worker("abacus.goal.deleted", Goal, :deleted),
      subscriber_worker("abacus.goal.updated", Goal, :updated),
      subscriber_worker("abacus.holding.created", Holding, :created),
      subscriber_worker("abacus.holding.deleted", Holding, :deleted),
      subscriber_worker("abacus.holding.updated", Holding, :updated),
      subscriber_worker("persona.member.created", Member, :created),
      subscriber_worker("persona.member.deleted", Member, :deleted),
      subscriber_worker("persona.member.updated", Member, :updated),
    ]
    supervise(children, strategy: :one_for_one, max_restarts: 10, max_seconds: 5)
  end

  def subscriber_worker(routing_key, module, function, opts \\ []) do
    prefetch = Keyword.get(opts, :prefetch, 2)
    queue_name = "brokaw.#{routing_key}"
    id = String.to_atom(queue_name)
    subscriber_config = %{
      queue: queue_name,
      prefetch: prefetch,
      consume_fn: at_least_once_consumer(module, function),
      setup_fn: setup_queue(queue_name, routing_key),
    }
    worker(Fireworks.Consumer, [subscriber_config, [name: id]], id: id)
  end

  def setup_queue(queue, routing_key) do
    fn(chan) ->
      alias AMQP.{Exchange, Queue}
      Exchange.topic(chan, "events", durable: false)
      Queue.declare(chan, queue, durable: false)
      Queue.bind(chan, queue, "events", routing_key: routing_key)
    end
  end

  # generates a consumer function that will call a given subscriber module
  defp at_least_once_consumer(subscriber_module, subscriber_fn) do
    fn(message, meta) ->
      require Logger
      Logger.info "Received #{meta.routing_key} (#{meta.delivery_tag})"
      %{payload: payload} = Exmx.EventWrapper.decode_event(message)
      apply(subscriber_module, subscriber_fn, [payload])
      Fireworks.Consumer.ack(meta)
    end
  end
end
```
</details>

TL/DR; I have a `subscriber_worker` helper function that creates my consumer and setup functions, and then I can easily setup a bunch of different consumers that route their messages to modules and functions in my application.

ewitchin commented:
@mmmries I didn't forget about it, got the flu and trying to catch up on work at LiveHelpNow. I will probably get to this early-mid next week. Thanks for the continued work!

ewitchin commented:
@mmmries I am going to spin up a test project and begin testing this now. Should probably have some good testing done by the end of the week. Looks very promising and a great update.

ewitchin commented:
Sorry for the delay in time it took testing and merging this.
